### PR TITLE
Define 'libs' property for vtk-m and vtk-h

### PR DESCRIPTION
**Pull request by scheibelp**
With this, dependent packages can easily find provided libraries via the Spack API (e.g. `spec['vtk-h'].libs`)

### CODE REVIEW

**Review by alalazo — DISMISSED** [review]
Waiting for maintainers to review before merging

File: var/spack/repos/builtin/packages/vtk-m/package.py
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py]: - We are missing: `vtkmdiympi(_nompi)` and other potential libraries that are generated when certain options are enabled.
- We need this to be sorted

**Comment by vicentebolea** [comment]
@scheibelp thanks for your contribution.

I am somewhat against this change for the VTK-m package since depending of the options we generate certain different libraries. While we can possibly check for certain variants to infer which library we will produce, this will also impose us to keep a strong dependency between the VTK-m source code changes with the Spack Package.

However, I can see that this is a good to have thing, thus a good compromise would be to only expose the very core libraries of VTK-m which at my judgement would be:

```
           "vtkm_rendering", -- if rendering variant is enabled
            "vtkm_filter_contour",
            "vtkm_filter_gradient",
            "vtkm_filter_extra",
            "vtkm_source",
            "vtkm_io",
            "vtkm_filter_common",
            "vtkm_cont",
            "vtkm_worklet",
```

Last thing is to split the changes into two commits, one for `vtk-h` and one for `vtk-m`

**Comment by scheibelp** [comment]
> I am somewhat against this change for the VTK-m package since depending of the options we generate certain different libraries. While we can possibly check for certain variants to infer which library we will produce, this will also impose us to keep a strong dependency between the VTK-m source code changes with the Spack Package.

FWIW Spack is OK with not finding some libraries, so basically you can list every library that this package could possibly produce, and then Spack will just return whatever ones happened to be installed (for whatever combination of variants are chosen for the particular `vtk-m` installation).

If that works for you then, I'd want to add `vtkmdiympi(_nompi)` (and you mentioned there were other libraries that could be installed). And I can also split this out into individual changes for `vtk-m` and `vtk-h` (I'll keep the `vtk-m` changes here).

**Comment by scheibelp** [comment]
> (I'll keep the vtk-m changes here).

I screwed up and kept the vtk-h changes here and put the `vtk-m` changes in https://github.com/spack/spack/pull/28065. Although you can answer/comment in either thread and I'll update as appropriate.

**Comment by cyrush** [comment]
@scheibelp  -- I agree with @vicentebolea. This is a tricky one.

There are lots of things that can change what libs are built (new versions of the project, options set, etc)
but it's even more nuanced than that. 

From our exported make logic in ascent, you can infer the libs we are currently tracking for vtk-m and vtk-h for folks using ascent downstream w/o cmake : 

https://github.com/Alpine-DAV/ascent/blob/abf8c5a22d269c73309e1015acd0d2cbeaea2823/src/config/ascent_config.mk.in#L127

One take away from this logic: Yes its mostly lib names and includes dirs, but there are choices downstream projects need to make. Providing everything, even when things missing are stripped out, won't work. 

Concrete  example:

Projects provide both both mpi and non mpi libs.  In some cases mpi libs can be used along with serial ones to add more capabilities (conduit is an example of this), but in other cases the mpi version truly supersedes the non-mpi (ascent is an example). Linking both is not the intended case and will result in disaster.

I feel these choices need to be made in the downstream build system when they leverage our products. 

We work to provide exported cmake targets and exported make logic (in the case of conduit, ascent, and i think now VTK-m) to help with this. This is logic is tested in our CI, including against new versions of TPLS during our update process. 

I would prefer folks use cmake imported targets (which are customized to solve problems like concrete one above ) or use our exported make targets (well, those are supported by ascent.mk -- not these libs themselves).

I worry spack lists will stagnate and increase surface error for what can go wrong.

**Comment by scheibelp** [comment]
> One take away from this logic: Yes its mostly lib names and includes dirs, but there are choices downstream projects need to make. Providing everything, even when things missing are stripped out, won't work.

Thanks for clarifying @cyrush. I think I could parameterize the `.libs` call to allow users to request the `_mpi` libs or not (`hdf5` has similar logic).

If packages should be consistent in which libraries to link (e.g. if all dependents should agree on whether to link with `_mpi...`) then I think this is another case of https://github.com/spack/seps/pull/2; in other cases like that, a temporary solution has just been to choose a consistent default subset for `libs` to return (although that might be inconsistent with what a build system chooses, typically packages don't invoke `libs` on a dependency if the build system already handles library selection for them).

**Comment by cyrush** [comment]
There are also different choices if CUDA is in the mix. 

If folks need to use these products in a new package, there is already work needed in the form of code to leverage their APIs. It seems the best path is for them to learn how to properly uses the projects. 

I understand the goal is to make this easy, but that means the spack recipe will need to duplicate and track logic that we already test and maintain in cmake. What is the value add vs maintenance burden?

**Comment by scheibelp** [comment]
> I understand the goal is to make this easy, but that means the spack recipe will need to duplicate and track logic that we already test and maintain in cmake. What is the value add vs maintenance burden?

(I should mention here that I'm not in a rush to merge these - I can maintain a separate repository for now - but I think this is a worthwhile thing to sort out so I'll keep probing)

* One benefit: Users don't have to be aware of CMake to use the appropriate libraries from this package (so they could write software that builds with tools other than CMake)
* https://github.com/spack/spack/pull/28061#issuecomment-997026228 outlines a scenario which, if true for this package, means that we would need to update it to make sure that there aren't linking errors: if packages need to agree on linking the `_mpi` libraries or not, then Spack needs to be made aware of this (this would apply regardless of whether all packages involved used CMake or not).

I'm curious if the library names/sets change often, because if not, then I don't think this would be a major maintenance issue (but let me know if you disagree). 

> There are also different choices if CUDA is in the mix.

I don't think that would add too much complexity:

```
def libs(self):
  if self.spec.satisfies(+cuda):
    # return one set of libs
  elif self.spec.satisfies(+mpi):
    # return a different set of libs
  else:
    ...
```

**Comment by cyrush** [comment]
I think the answer is that you need to effectively re-create the logic that we already provide for makefile based systems.

These are the examples:
https://github.com/LLNL/conduit/blob/develop/src/config/conduit_config.mk.in
https://github.com/Alpine-DAV/ascent/blob/develop/src/config/ascent_config.mk.in

For ascent + cuda (for device linking reasons) -- there is an extra object file instead of an extra lib.

For what it's worth, the reason why our vtk-h and vtk-m makefile stuff is in ascent (and instead of in, say, vtk-h), is b/c the ascent build system has bottled it up and completed the complex cuda device linking steps necessary to use vtk-h and vtk-m more easily downstream. Even for codes that are not using cuda. This is b/c we don't want to require downstream codes to have to link with cuda, if they aren't using it.  I think that might also be why this is a sore point and one that is difficult to convey. This process was painful and absolutely requires cmake to get it right.

CMake really is the preferred path and the backup is our exported makefile logic .
We test both as we develop our code, and we try to make it easy for folks to use these paths in our docs:
https://ascent.readthedocs.io/en/latest/QuickStart.html#using-ascent-in-your-project

I guess it's unlikely that folks will not leverage existing paths - so maybe this is not a big issue.  

But spack package maintenance is already non trivial and it worries me to add things that will be difficult to test and likely to drift.

**Comment by vicentebolea** [comment]
I agree with @scheibelp that this can benefit potential users which wish to use VTK-m without using CMake, which at this point of time could be achieved by re-creating in the Spack package those couple of conditionals referred by @cyrush.

On the other hand, I am hesitant to support or ease, at the expense of potential additional burden in the future, users not using CMake. The reason is that there seems not to be a sizeable amount of users who demands this.

I opt for keeping this thread open, and see if there is interest in supporting this, if so we can commit to keep this in sync with the VTK-m master.

> I'm curious if the library names/sets change often

For current ones not really, but we keep supporting more parallel computing platforms/library which could potentially include more libraries.

**Comment by alalazo** [comment]
@scheibelp Is this still a work in progress, or can we close the PR?

**Comment by alalazo** [comment]
@scheibelp Going through the list of PRs that are assigned to me. I'm going to close this, but feel free to reopen if considered still a work in progress.

**Pull request closed without merging**
